### PR TITLE
Upload new files for a sharing

### PR DESCRIPTION
**Pull request by nono**

### CODE REVIEW

File: pkg/sharing/replicator.go [hunk 1]
```diff
@@ -237,6 +237,9 @@ func (s *Sharing) callChangesFeed(inst *instance.Instance, since string) (*Chang
 		if !ok {
 			continue
 		}
+		if _, ok = info["binary"]; ok {
+			continue
```
Comment (Gara64) [inline pkg/sharing/replicator.go]: Shouldn't it be `!ok` ?

File: pkg/workers/share/share.go [hunk 2]
```diff
@@ -23,20 +32,42 @@ func init() {
...
-		MaxExecCount: 2,
-		Timeout:      30 * time.Second,
-		WorkerFunc:   WorkerTrack,
+		WorkerType:  "share-upload",
+		Concurrency: runtime.NumCPU(),
+		// TODO the worker is not idempotent: if it fails, it adds a new job to
+		// retry, but with MaxExecCount > 1, it can amplifies a lot the number
+		// of retries
+		MaxExecCount: 1,
+		Timeout:      1 * time.Hour,
```
Comment (Gara64) [inline pkg/workers/share/share.go]: A large video might fail with a slow connection, but do we support this kind of case ?

File: pkg/sharing/setup.go [hunk 3]
```diff
@@ -84,7 +84,13 @@ func (s *Sharing) Setup(inst *instance.Instance, m *Member) {
 	if err := s.ReplicateTo(inst, m, true); err != nil {
 		inst.Logger().WithField("nspace", "sharing").
 			Warnf("Error on initial replication (%s): %s", s.SID, err)
-		s.retryReplicate(inst, 0)
+		s.retryWorker(inst, "share-replicate", 0)
+	} else {
+		if err := s.InitialUpload(inst, m); err != nil {
```
Comment (Gara64) [inline pkg/sharing/setup.go]: We should check the sharing involves files here

File: pkg/sharing/upload.go [hunk 4]
```diff
@@ -95,11 +98,63 @@ func (s *Sharing) UploadTo(inst *instance.Instance, m *Member) (bool, error) {
...
 	}
 
 	return true, s.UpdateLastSequenceNumber(inst, m, "upload", seq)
 }
+
+// findNextFileToUpload uses the changes feed to find the next file that needs
+// to be uploaded. It returns a file document if there is one file to upload,
+// and the sequence number where it is in the changes feed.
+func (s *Sharing) findNextFileToUpload(inst *instance.Instance, since string) (*vfs.FileDoc, string, error) {
+	for {
```
Comment (Gara64) [inline pkg/sharing/upload.go]: Isn't a bit dangerous? Maybe a MAX_CHANGES condition could be added

File: docs/sharing.md [hunk 5]
```diff
@@ -561,4 +561,61 @@ Content-Type: application/json
...
+PUT /sharings/ce8835a061d0ef68947afe69a0046722/io.cozy.files/0c1116b028c6ae6f5cdafb949c088265/metadata HTTP/1.1
+Host: bob.example.net
+Accept: application/json
+Content-Type: application/json
+Authorization: Bearer ...
+```
+
+#### Response
+
+If only the metadata has changed (not the content), the response will be a
```
Comment (Gara64) [inline docs/sharing.md]: If only -> Only if :)

File: pkg/sharing/upload.go [hunk 6]
```diff
@@ -305,3 +309,32 @@ func (s *Sharing) uploadFile(inst *instance.Instance, m *Member, file *couchdb.J
...
+// HandleFileUpload is used to receive a file upload when synchronizing just
+// the metadata was not enough.
+func (s *Sharing) HandleFileUpload(inst *instance.Instance, key string, body io.ReadCloser) error {
+	newdoc := getCache().Get(inst.Domain, key)
+	if newdoc == nil {
+		return ErrMissingFileMetadata
+	}
+
+	indexer := NewSharingIndexer(inst, nil) // TODO
+	fs := inst.VFS().UseSharingIndexer(indexer)
```
Comment (Gara64) [inline pkg/sharing/upload.go]: Why do we need a special indexer for the sharings?

File: pkg/sharing/replicator.go (hunk 1 above)
Comment (nono) [reply pkg/sharing/replicator.go]: No, I really meant `ok`. We are here in the replicator (the first phase of synchronization of a sharing), and binaries are managed in the second phase ("the upload"), so we want to skip them here.

File: pkg/workers/share/share.go (hunk 2 above)
Comment (nono) [reply pkg/workers/share/share.go]: We will support this kind of case, but latter. For the moment, the limit of 1 hour seems to be a good tradeoff.

File: pkg/sharing/upload.go (hunk 4 above)
Comment (nono) [reply pkg/sharing/upload.go]: Well, maybe we can

File: docs/sharing.md (hunk 5 above)
Comment (nono) [reply docs/sharing.md]: What I wanted to say is that if the metadata has changed, the response will be a 204, but if the metadata and the content (or just the content) have changed, the response will be a 200. In short: if "only the metadata has changed", then 204.

File: pkg/sharing/upload.go [hunk 7]
```diff
@@ -166,6 +167,43 @@ func (s *Sharing) findNextFileToUpload(inst *instance.Instance, since string) (*
...
+type KeyToUpload struct {
+	Key string `json:"key"`
+}
+
+func (s *Sharing) createUploadKey(inst *instance.Instance, target *vfs.FileDoc) (*KeyToUpload, error) {
+	key := getCache().Save(inst.Domain, target)
+	return &KeyToUpload{Key: key}, nil
+}
+
+// SyncFile tries to synchroniza a file with just the metadata. If it can't,
```
[the remark this answers is not in the document]
Comment (nono) [reply pkg/sharing/upload.go]: Good catch!

File: pkg/sharing/upload.go [hunk 8]
```diff
@@ -245,12 +283,11 @@ func (s *Sharing) uploadFile(inst *instance.Instance, m *Member, file *couchdb.J
...
 	defer content.Close()
 
-	// TODO send resBody["token"]
 	res2, err := request.Req(&request.Options{
 		Method: http.MethodPut,
 		Scheme: u.Scheme,
 		Domain: u.Host,
-		Path:   "/sharings/" + s.SID + "/io.cozy.files/" + file.ID(),
+		Path:   "/sharings/" + s.SID + "/io.cozy.files/" + resBody.Key,
 		Headers: request.Headers{
```
[the remark this answers is not in the document]
Comment (nono) [reply pkg/sharing/upload.go]: I use the status code to know if we need to upload the binary or not.

**Pull request merged**